File: projects/frontend/src/types/order.ts
export const OrderStatus = {
  Draft: 0,
  WaitingDesign: 1,
  InDesign: 2,
  WaitingApproval: 3,
  Approved: 4,
  Rejected: 5,
  InProduction: 6,
  Shipped: 7,
} as const;
export type OrderStatus = (typeof OrderStatus)[keyof typeof OrderStatus];

export const DesignRequestType = {
  CustomerUpload: 0,
  RequestFromUs: 1,
  Template: 2,
} as const;
export type DesignRequestType = (typeof DesignRequestType)[keyof typeof DesignRequestType];

export const DesignRequestStatus = {
  WaitingUpload: 0,
  CustomerUploaded: 1,
  PrintRejected: 2,
  PrintApproved: 3,
  InDesign: 4,
  WaitingApproval: 5,
  Approved: 6,
  Rejected: 7,
} as const;
export type DesignRequestStatus = (typeof DesignRequestStatus)[keyof typeof DesignRequestStatus];

export const DeliveryType = {
  Tracked48: 0,
  NextDay: 1,
} as const;
export type DeliveryType = (typeof DeliveryType)[keyof typeof DeliveryType];

export interface OrderListItem {
  id: string;
  ebayOrderNo: string;
  customerName: string;
  customerEmail: string;
  status: OrderStatus;
  itemCount: number;
  createdDate: string;
}

export interface OrderDetail {
  id: string;
  ebayOrderNo: string;
  status: OrderStatus;
  notes: string | null;
  createdDate: string;
  customer: CustomerSummary;
  items: OrderItem[];
  designRequests: DesignRequestSummary[];
  shipment: ShipmentSummary | null;
}

export interface CustomerSummary {
  id: string;
  customerName: string;
  email: string;
}

export interface OrderItem {
  id: string;
  sku: string;
  quantity: number;
  description: string | null;
}

export interface DesignRequestSummary {
  id: string;
  orderItemId: string;
  type: DesignRequestType;
  status: DesignRequestStatus;
  fileCount: number;
}

export interface ShipmentSummary {
  id: string;
  trackingNumber: string;
  carrier: string;
  deliveryType: DeliveryType;
  shipmentDate: string;
}

export interface CreateOrderRequest {
  ebayOrderNo: string;
  customerId: string;
  notes?: string;
  items: CreateOrderItemRequest[];
}

export interface CreateOrderItemRequest {
  sku: string;
  quantity: number;
  description?: string;
  designType: DesignRequestType;
}

File: projects/backend/src/EbayDesign.Domain/Entities/Order.cs
using EbayDesign.Domain.Common;
using EbayDesign.Domain.Enums;
using EbayDesign.Domain.Events;

namespace EbayDesign.Domain.Entities;

public class Order : BaseEntity
{
    public Guid CustomerId { get; set; }
    public string EbayOrderNo { get; set; } = string.Empty;
    public OrderStatus Status { get; set; } = OrderStatus.Draft;
    public string? Notes { get; set; }

    public Customer Customer { get; set; } = null!;
    public ICollection<OrderItem> Items { get; set; } = new List<OrderItem>();
    public ICollection<DesignRequest> DesignRequests { get; set; } = new List<DesignRequest>();
    public Shipment? Shipment { get; set; }

    private static readonly Dictionary<(OrderStatus From, OrderStatus To), bool> _allowedTransitions = new()
    {
        { (OrderStatus.Draft, OrderStatus.WaitingDesign), true },
        { (OrderStatus.WaitingDesign, OrderStatus.InDesign), true },
        { (OrderStatus.InDesign, OrderStatus.WaitingApproval), true },
        { (OrderStatus.WaitingApproval, OrderStatus.Approved), true },
        { (OrderStatus.WaitingApproval, OrderStatus.Rejected), true },
        { (OrderStatus.Rejected, OrderStatus.InDesign), true },
        { (OrderStatus.Approved, OrderStatus.InProduction), true },
        { (OrderStatus.InProduction, OrderStatus.Shipped), true },
    };

    public void TransitionTo(OrderStatus newStatus)
    {
        if (!_allowedTransitions.ContainsKey((Status, newStatus)))
        {
            throw new InvalidOperationException(
                $"Cannot transition order from '{Status}' to '{newStatus}'.");
        }

        var oldStatus = Status;
        Status = newStatus;

        AddDomainEvent(new OrderStatusChangedEvent(Id, oldStatus, newStatus));

        if (newStatus == OrderStatus.Draft)
            AddDomainEvent(new OrderCreatedEvent(Id));
    }
}
